File: Algo/Leetcode/055JumpGame.cpp
// 从后往前dp。会超时O(nk)
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        if(n==0) return true;
        vector<bool> dp(n, false);
        dp[n-1] = true;
        for(int i=n-2; i>=0; i--){
            for(int j=1; j<=nums[i]; j++){
                if(i+j<nums.size()) dp[i] = (dp[i] || dp[i+j]);
            }
        }
        return dp[0];
    }
};

// 从前往后dp，只要贪心记录当前最右能够到的位置即可
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        if(n==0) return true;
        int r = 0;
        for(int i=0; i<n; i++){
            if(i<=r) r = max(r, i+nums[i]);
            if(r >= n-1) return true;
        }
        return false;
    }
};

File: Algo/Leetcode/055JumpGame.py
class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        max_len = 0     # update max len as the most far idx currently can access
        l = len(nums)
        for i, n in enumerate(nums):
            if max_len >= l-1:
                return True
            if max_len < i:
                return False
            max_len = max(max_len, i+n)

        return True